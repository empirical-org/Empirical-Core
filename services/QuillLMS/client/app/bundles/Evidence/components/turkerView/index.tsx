import * as React from "react";
import { connect } from "react-redux";

import TurkLanding from './landing';
import TurkCompleted from './completed';

import StudentView from '../studentView/container';

export const TurkerView = ({ session }) => {
  const { sessionID } = session;
  const [activityStarted, setActivityStarted] = React.useState<boolean>(false);
  const [activityCompleted, setActivityCompleted] = React.useState<boolean>(false);

  function handleStartActivity() {
    setActivityStarted(true);
  }

  function handleFinishActivity() {
    setActivityCompleted(true);
  }

  function componentToShow() {
    if (!activityStarted) { return <TurkLanding handleStartActivity={handleStartActivity} /> }

    if (activityStarted && !activityCompleted) { return <StudentView handleFinishActivity={handleFinishActivity} isTurk={true} />}

    return <TurkCompleted code={sessionID} />
 }

  return(
    <div className="turker-view-container">
      {componentToShow()}
    </div>
  )
}

const mapStateToProps = (state: any) => {
  return {
    session: state.session
  };
};

export default connect(mapStateToProps, {})(TurkerView);
