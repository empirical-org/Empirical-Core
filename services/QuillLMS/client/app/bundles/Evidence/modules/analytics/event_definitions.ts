class Event {
  name: string;
  requiredProperties?: Array<string>;

  constructor(name: string, requiredProperties?: Array<string>) {
    this.name = name;
    this.requiredProperties = requiredProperties;
  }
}


// Even though we ultimately want to expose a single object with
// different events differentiated by key, we define our events
// as an array of simple objects so that we can validate that
// as the list gets longer we don't accidentally end up with a
// name collision (see validation in the 'events.ts' file).
const EventDefinitions = [
  {COMPREHENSION_ACTIVITY_STARTED: new Event('comprehensionActivityStarted', [
    'activityID',
    'sessionID'])},
  {COMPREHENSION_PASSAGE_READ: new Event('comprehensionPassageRead', [
    'activityID',
    'sessionID'])},
  {COMPREHENSION_PROMPT_STARTED: new Event('comprehensionPromptStarted', [
    'activityID',
    'promptID',
    'sessionID'])},
  {COMPREHENSION_ENTRY_SUBMITTED: new Event('comprehensionEntrySubmitted', [
    'activityID',
    'attemptNumber',
    'promptID',
    'promptStemText',
    'sessionID',
    'startingFeedback',
    'startingFeedbackID',
    'submittedEntry'])},
  {COMPREHENSION_FEEDBACK_RECEIVED: new Event('comprehensionFeedbackReceived', [
    'activityID',
    'attemptNumber',
    'promptID',
    'promptStemText',
    'returnedFeedback',
    'returnedFeedbackID',
    'sessionID',
    'startingFeedback',
    'startingFeedbackID',
    'submittedEntry'])},
  {COMPREHENSION_PROMPT_COMPLETED: new Event('comprehensionPromptCompleted', [
    'activityID',
    'promptID',
    'sessionID'])},
  {COMPREHENSION_ACTIVITY_COMPLETED: new Event('comprehensionActivityCompleted', [
    'activityID',
    'sessionID'])},
  {COMPREHENSION_ACTIVITY_SAVED: new Event('comprehensionActivitySaved', [
    'activityID',
    'sessionID'])}
];


export {
  Event,
  EventDefinitions,
};
