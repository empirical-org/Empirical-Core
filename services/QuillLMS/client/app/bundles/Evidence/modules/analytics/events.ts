import { Event, EventDefinitions } from './event_definitions';


// Ensure that no two events accidentally share the same name,
// either as the constant identifier, or as the literal passed
// to analytics
function validateEventsList(eventDefinitions: Array<any>): void {
  function checkForDuplicates(target: Array<any>) {
    if (new Set(target).size != target.length) {
      throw new Error('There are duplicate values in the provided array.');
    }
  }
  let eventReferences = eventDefinitions.map((d) => Object.keys(d)[0]);
  let eventNames = eventDefinitions.map((d) => Object.values(d)[0]['name']);
  try {
    checkForDuplicates(eventReferences);
    checkForDuplicates(eventNames);
  } catch {
    throw new Error('At least two defined events have the same name assigned.');
  }
}
validateEventsList(EventDefinitions);


// Turn our array of definitions into a nice, referenceable object
function arrayToObject(sourceArray) {
  return sourceArray.reduce((accumulator, eventDefinition) => Object.assign(accumulator, eventDefinition), {});
}

let buildEvents = arrayToObject(EventDefinitions)
buildEvents.providers = {
  CLEVER: 'clever',
  EMAIL: 'email',
  GOOGLE: 'google',
};
const Events = buildEvents;


export default Events;
