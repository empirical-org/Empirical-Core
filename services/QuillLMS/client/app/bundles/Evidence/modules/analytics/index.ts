import { Event } from './event_definitions';
import Events from './events';


class SegmentAnalytics {
  analytics(): object {
    return window.analytics
  }

  track(event: Event, properties?: object): boolean {
    try {
      // Make sure that the event reference is one that's defined
      if (!event) {
        throw new Error('The event referenced does not exist.');
      }

      // Validate that required properties are present
      this.validateEvent(event, properties);

      // Check to make sure that we have access to the analytics global
      if (!this.analytics()) {
        throw new Error(`Error sending event '${event.name}'.  SegmentAnalytics was instantiated before an instance of window.analytics could be found to connect to.`);
      }
    } catch(e) {
      this.reportError(e);
      return false;
    }

    const eventProperties = Object.assign(this.formatCustomProperties(properties), this.getDefaultProperties());

    this.analytics().track(event.name, eventProperties);
    return true;
  }

  validateEvent (event: Event, properties?: object): void {
    if (properties === undefined) {
      properties = {};
    }
    if (event.requiredProperties) {
      let passedEventProperties = Object.keys(properties);
      event.requiredProperties.forEach((p) => {
        if (passedEventProperties.indexOf(p) == -1) {
          throw new Error(`Can not track event "${event.name}" without required property "${p}".`);
        }
      });
    }
  }

  formatCustomProperties(properties: object): object {
    if (typeof properties != 'object') {
      properties = {};
    }
    return Object.keys(properties).reduce((accumulator, key) => {
      let customKeyName = `custom_${key}`;
      accumulator[customKeyName] = properties[key];
      return accumulator;
    }, {});
  }

  getDefaultProperties(): object {
    return {
      path: `${window.location.pathname}${window.location.search}${window.location.hash}`,
      referrer: document.referrer,
    };
  }

  reportError(e: Error): void {
    // placeholder for actual error reporting
    // to do - do something with this error
  }
}

const segmentInstance = new SegmentAnalytics();

export {
  segmentInstance as SegmentAnalytics,
  Events,
};
