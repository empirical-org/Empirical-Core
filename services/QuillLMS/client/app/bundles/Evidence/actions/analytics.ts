import * as request from 'request';

import { Event } from '../modules/analytics/event_definitions'
import { SegmentAnalytics } from '../modules/analytics'

export const TrackAnalyticsEvent = (event: Event, params: Object) => {
  return (dispatch: Function) => {
    SegmentAnalytics.track(event, params);
  }
}
