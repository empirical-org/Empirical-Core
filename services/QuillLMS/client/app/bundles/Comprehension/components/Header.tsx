import * as React from "react";
import * as Redux from "redux";
import { connect } from "react-redux";
import { TrackAnalyticsEvent } from '../actions/analytics';
import { Events } from '../modules/analytics';
import '../styles/headerStyling.scss'

import getParameterByName from '../helpers/getParameterByName';

const logoSrc = `${process.env.CDN_URL}/images/logos/quill-logo-white.svg`
const mobileLogoSrc = `${process.env.CDN_URL}/images/logos/quill-logo-white-mobile.svg`

export class Header extends React.Component<any, any> {
  constructor(props: any) {
    super(props)

    this.state = {
      sessionID: getParameterByName('student', window.location.href)
    }
  }

  goToLMS = () => {
    window.location.href = "/"
  }

  trackSaveAndExitEvent = () => {
    const { dispatch, } = this.props
    const { sessionID, } = this.state
    const activityID = getParameterByName('uid', window.location.href)
    dispatch(TrackAnalyticsEvent(Events.COMPREHENSION_ACTIVITY_SAVED, {
      activityID: activityID,
      sessionID,
    }))
  }

  saveAndExit = () => {
    const { sessionID, } = this.state
    this.trackSaveAndExitEvent()
    if (sessionID) {
    } else {
      this.goToLMS()
    }
  }

  handleOnClick = () => {
    this.saveAndExit()
  }

  render() {
    const isNotTurk = !window.location.href.includes('turk')
    return (
      <div className="header">
        <div>
          <img alt="Quill.org logo" className="hide-on-desktop" src={mobileLogoSrc} />
          <img alt="Quill.org logo" className="hide-on-mobile" src={logoSrc} />
          {isNotTurk && <button className="save-and-exit focus-on-dark" onClick={this.handleOnClick} type="button"><span>Save and exit</span></button>}
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state: any) => {
  return {
    session: state.session
  };
};

const mapDispatchToProps = (dispatch: Redux.Dispatch<any>) => {
  return {
      dispatch
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header)
