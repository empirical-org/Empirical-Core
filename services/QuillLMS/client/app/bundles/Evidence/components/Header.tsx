import * as React from "react";
import * as Redux from "redux";
import { connect } from "react-redux";
import { TrackAnalyticsEvent } from '../actions/analytics';
import { Events } from '../modules/analytics';
import '../styles/headerStyling.scss'

import getParameterByName from '../helpers/getParameterByName';
import { Tooltip, READ_PASSAGE_STEP_NUMBER, BECAUSE_PASSAGE_STEP_NUMBER, BUT_PASSAGE_STEP_NUMBER, SO_PASSAGE_STEP_NUMBER, whiteCheckGreenBackgroundIcon } from '../../Shared/index'
import { onMobile } from '../helpers/containerActionHelpers';

const logoSrc = `${process.env.CDN_URL}/images/logos/quill-logo-white-2022.svg`
const mobileLogoSrc = `${process.env.CDN_URL}/images/logos/quill-logo-white-2022-mobile.svg`
const helpIcon = `${process.env.CDN_URL}/images/icons/icons-help-white.svg`
const checkIcon = <img alt={whiteCheckGreenBackgroundIcon.alt} src={whiteCheckGreenBackgroundIcon.src} />;

export class Header extends React.Component<any, any> {
  constructor(props: any) {
    super(props)

    this.state = {
      sessionID: getParameterByName('student', window.location.href)
    }
  }

  goToLMS = () => {
    window.location.href = "/"
  }

  trackSaveAndExitEvent = () => {
    const { dispatch } = this.props
    const { sessionID, } = this.state
    const activityID = getParameterByName('uid', window.location.href)
    dispatch(TrackAnalyticsEvent(Events.EVIDENCE_ACTIVITY_SAVED, {
      activityID: activityID,
      sessionID
    }))
  }

  saveAndExit = () => {
    const { sessionID, } = this.state
    this.trackSaveAndExitEvent()
    if (sessionID) {
    } else {
      this.goToLMS()
    }
  }

  handleOnClick = () => {
    this.saveAndExit()
  }

  renderNumberOrIcon = (step: number) => {
    const { session } = this.props;
    const { activeStep } = session;
    if(step < activeStep) {
      return checkIcon;
    }
    if(step === activeStep) {
      return <div className="evidence-step-number-small active">{step}</div>
    }
    return <div className="evidence-step-number-small">{step}</div>
  }

  renderStepCounter = () => {
    return(
      <div className="nav-steps-count-container">
        <p className="step-counter-label">Steps</p>
        {this.renderNumberOrIcon(READ_PASSAGE_STEP_NUMBER)}
        {this.renderNumberOrIcon(BECAUSE_PASSAGE_STEP_NUMBER)}
        {this.renderNumberOrIcon(BUT_PASSAGE_STEP_NUMBER)}
        {this.renderNumberOrIcon(SO_PASSAGE_STEP_NUMBER)}
      </div>
    );
  }

  render() {
    const { session } = this.props;
    const { explanationSlidesCompleted, activityIsComplete } = session;
    const showStepsCounter = explanationSlidesCompleted && !activityIsComplete
    const isNotTurk = !window.location.href.includes('turk')
    const tooltipTrigger = <div><img alt="Question mark icon" src={helpIcon} /><span>Beta: <span>in development</span></span></div>
    const mobileStyle = onMobile() || !showStepsCounter ? 'mobile' : '';
    return (
      <div className="header">
        <div className={`inner-header ${mobileStyle}`}>
          <div className={`left-side-container ${mobileStyle}`}>
            <div className="left-side">
              <a className="focus-on-dark" href={process.env.DEFAULT_URL}>
                <img alt="Quill.org logo" className="hide-on-desktop" src={mobileLogoSrc} />
                <img alt="Quill.org logo" className="hide-on-mobile" src={logoSrc} />
              </a>
              <Tooltip
                tooltipText="Quill Reading for Evidence is in beta, which means that it’s not perfect yet. As you complete activities, you may notice some issues—a button may not work or some feedback could be unhelpful. Please know that we’re actively working on Quill Reading for Evidence and making improvements every day."
                tooltipTriggerText={tooltipTrigger}
                tooltipTriggerTextClass="hide-on-mobile beta-tag focus-on-dark"
              />
              <span className="hide-on-desktop beta-tag">Beta</span>
            </div>
          </div>
          <div className={`right-side-container ${mobileStyle}`}>
            {showStepsCounter && !onMobile() && this.renderStepCounter()}
            {isNotTurk && <button className="save-and-exit focus-on-dark" onClick={this.handleOnClick} type="button"><span>Save and exit</span></button>}
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state: any) => {
  return {
    session: state.session
  };
};

const mapDispatchToProps = (dispatch: Redux.Dispatch<any>) => {
  return {
    dispatch
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header)
