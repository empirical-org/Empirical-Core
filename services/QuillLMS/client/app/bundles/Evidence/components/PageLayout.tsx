import * as React from "react";
import { renderRoutes } from "react-router-config";
import { useQuery } from 'react-query';

import Header from "./Header";

import { fetchUserRole } from '../../Shared/utils/userAPIs';
import { ScreenreaderInstructions, TeacherPreviewMenu, INTRODUCTION } from '../../Shared/index';
import { routes } from "../routes";
import getParameterByName from '../helpers/getParameterByName';

const PageLayout = (props: any) => {
  const { user } = props;
  const studentSession = getParameterByName('student', window.location.href);
  const turkSession = getParameterByName('turk', window.location.href);
  const isPlaying = window.location.href.includes('play');
  const studentOrTurk = studentSession || turkSession;
  const { data } = useQuery("user-role", fetchUserRole);
  const isTeacherOrAdmin = data && data.role && data.role !== 'student';

  const [previewShowing, setPreviewShowing] = React.useState<boolean>(!studentOrTurk);
  const [questionToPreview, setQuestionToPreview] = React.useState<any>(INTRODUCTION);

  function handleSkipToMainContentClick () {
    const element = document.getElementById("main-content")
    if (!element) { return }
    element.focus()
    element.scrollIntoView()
  }

  function handleTogglePreviewMenu() {
    setPreviewShowing(!previewShowing);
  }

  function handleToggleQuestion(step: string) {
    setQuestionToPreview(step);
  }

  const showPreview = previewShowing && isTeacherOrAdmin && isPlaying;
  const isOnMobile = window.innerWidth < 1100;

  return (
    <div className="app-container">
      <ScreenreaderInstructions />
      <button className="skip-main" onClick={handleSkipToMainContentClick} type="button">Skip to main content</button>
      <Header isOnMobile={isOnMobile} isTeacher={!studentOrTurk} onTogglePreview={handleTogglePreviewMenu} previewShowing={showPreview} />
      <div className="activity-container">
        {showPreview && <aside
          className="sider-container"
          style={{ overflowY: 'auto', width: '360px' }}
        >
          <TeacherPreviewMenu
            isOnMobile={isOnMobile}
            onTogglePreview={handleTogglePreviewMenu}
            onToggleQuestion={handleToggleQuestion}
            questionToPreview={questionToPreview}
            showPreview={previewShowing}
          />
        </aside>}
        <div id="main-content" tabIndex={-1}>{renderRoutes(routes, {
          user,
          isOnMobile: isOnMobile,
          handleTogglePreviewMenu: handleTogglePreviewMenu,
          handleToggleQuestion: handleToggleQuestion,
          previewMode: showPreview,
          questionToPreview: questionToPreview
          })}</div>
      </div>
    </div>
  );
};

export default PageLayout;
