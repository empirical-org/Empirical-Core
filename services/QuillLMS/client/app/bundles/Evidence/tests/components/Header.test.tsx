import * as React from 'react'
import { mount, } from 'enzyme';
import toJson from 'enzyme-to-json';

const mockTrackAnalyticsEvent = jest.fn()
jest.mock('../../actions/analytics', () => ({
  TrackAnalyticsEvent: mockTrackAnalyticsEvent
}))

const MockSessionID = 'MockSessionId'
const MockActivityID = 1
const mockGetParameterByName = jest.fn()
  .mockReturnValueOnce(MockSessionID)
  .mockReturnValue(MockActivityID)
jest.mock('../../helpers/getParameterByName', () => ({
  default: mockGetParameterByName
}))

import { Header } from '../../components/Header'
import { Events } from '../../modules/analytics'

const dispatch = () => {}

describe('StudentViewContainer component', () => {
  describe('when the activity has loaded', () => {
    const wrapper = mount(<Header
      dispatch={dispatch}
    />)

    it('renders', () => {
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    describe('when a user clicks "Save and exit"', () => {
      it('should track a COMPREHENSION_ACTIVITY_SAVED event', () => {
        wrapper.find('.save-and-exit').simulate('click')

        expect(mockTrackAnalyticsEvent).toHaveBeenCalledWith(Events.COMPREHENSION_ACTIVITY_SAVED, {
          activityID: MockActivityID,
          sessionID: MockSessionID
        })
      })
    })
  })
})
