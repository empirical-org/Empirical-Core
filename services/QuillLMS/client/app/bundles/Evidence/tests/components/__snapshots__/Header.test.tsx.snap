// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentViewContainer component when the activity has loaded renders 1`] = `
<Header
  dispatch={[Function]}
  session={Object {}}
>
  <div
    className="header"
  >
    <div
      className="inner-header mobile"
    >
      <div
        className="left-side-container mobile"
      >
        <div
          className="left-side"
        >
          <a
            className="focus-on-dark"
          >
            <img
              alt="Quill.org logo"
              className="hide-on-desktop"
              src="undefined/images/logos/quill-logo-white-mobile.svg"
            />
            <img
              alt="Quill.org logo"
              className="hide-on-mobile"
              src="undefined/images/logos/quill-logo-white-2022.svg"
            />
          </a>
          <Tooltip
            isTabbable={true}
            tooltipText="Quill Reading for Evidence is in beta, which means that it’s not perfect yet. As you complete activities, you may notice some issues—a button may not work or some feedback could be unhelpful. Please know that we’re actively working on Quill Reading for Evidence and making improvements every day."
            tooltipTriggerText={
              <div>
                <img
                  alt="Question mark icon"
                  src="undefined/images/icons/icons-help-white.svg"
                />
                <span>
                  Beta: 
                  <span>
                    in development
                  </span>
                </span>
              </div>
            }
            tooltipTriggerTextClass="hide-on-mobile beta-tag focus-on-dark"
          >
            <span
              aria-hidden={false}
              className="quill-tooltip-trigger  "
              role="tooltip"
            >
              <span
                className="hide-on-mobile beta-tag focus-on-dark"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                role="button"
                tabIndex={0}
              >
                <div>
                  <img
                    alt="Question mark icon"
                    src="undefined/images/icons/icons-help-white.svg"
                  />
                  <span>
                    Beta: 
                    <span>
                      in development
                    </span>
                  </span>
                </div>
              </span>
              <span
                className="quill-tooltip-wrapper"
              >
                <span
                  aria-live="polite"
                  className="quill-tooltip"
                />
              </span>
            </span>
          </Tooltip>
          <span
            className="hide-on-desktop beta-tag"
          >
            Beta
          </span>
        </div>
      </div>
      <div
        className="right-side-container mobile"
      >
        <button
          className="save-and-exit focus-on-dark"
          onClick={[Function]}
          type="button"
        >
          <span>
            Save and exit
          </span>
        </button>
      </div>
    </div>
  </div>
</Header>
`;
